File: projectforge-business/src/test/java/org/projectforge/messaging/SmsSenderTest.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2014 Kai Reinhard (k.reinhard@micromata.de)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.messaging;

import org.apache.commons.httpclient.HttpClient;
import org.apache.commons.httpclient.HttpMethod;
import org.apache.commons.httpclient.methods.GetMethod;
import org.apache.commons.httpclient.methods.PostMethod;
import org.projectforge.sms.SmsSenderConfig;
import org.testng.Assert;
import org.testng.annotations.Test;
import org.yaml.snakeyaml.Yaml;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import static org.mockito.Mockito.*;

public class SmsSenderTest {
  private static final org.slf4j.Logger log = org.slf4j.LoggerFactory.getLogger(SmsSenderTest.class);

  @Test
  public void testSmsService() throws Exception {
    Assert.assertEquals(SmsSender.HttpResponseCode.SUCCESS,
            testGetCall("hurzel", "0123456", "Hello_world", 200, "OK. Perfect."));
    Assert.assertEquals(SmsSender.HttpResponseCode.UNKNOWN_ERROR,
            testGetCall("hurzel", "0123456", "Hello_world", 200, "ERROR"));
    Assert.assertEquals(SmsSender.HttpResponseCode.MESSAGE_TO_LARGE,
            testGetCall("hurzel", "0123456", "Hello_world", 200, "MESSAGE TO LARGE"));
    Assert.assertEquals(SmsSender.HttpResponseCode.UNKNOWN_ERROR,
            testPostCall("hurzel", "0123456", "Hello_world", 100, "Ingore this message."));
  }

  /**
   * Configuration of live test in ~/ProjectForge/smsTestConfig.yaml
   * <pre>
   *   projectforge:
   *     sms:
   *         httpMethod: POST
   *         httpParameters:
   *             mode: 'number'
   *             password: 'mypassword'
   *             text: '#message'
   *             to: '#number'
   *             username: 'projectforge'
   *         returnCodePattern:
   *             error: ''
   *             messageError: ''
   *             messageToLargeError: ''
   *             numberError: ''
   *             success: ''
   *         url: 'http://smsgateway.acme.com/api.php'
   *         testNumber: '0123456789'
   * </pre>
   *
   * @throws IOException
   */
  @Test
  public void liveTest() throws IOException {
    // ~/ProjectForge/smsTestConfig.yaml
    File configFile = new File(System.getProperty("user.home"), "ProjectForge" + File.separatorChar + "smsTestConfig.yaml");
    if (!configFile.exists()) {
      log.info("Skipping live testing of sms service (OK). For live testing, please refer this code for the required config file: " + configFile.getAbsolutePath());
      return;
    }
    Yaml yaml = new Yaml();
    Map<String, Object> map = yaml.load(new FileReader(configFile));
    String httpMethodType = (String) getFromMap(map, "projectforge", "sms", "httpMethod");
    String url = (String) getFromMap(map, "projectforge", "sms", "url");
    String testNumber = (String) getFromMap(map, "projectforge", "sms", "testNumber");
    Map<String, String> httpParams = (Map<String, String>) getFromMap(map, "projectforge", "sms", "httpParameters");
    SmsSenderConfig config = new SmsSenderConfig();
    config.setHttpMethodType(httpMethodType).setUrl(url);
    config.setHttpParams(httpParams);
    config.setSmsMaxMessageLength((int) getFromMap(map, "projectforge", "sms", "smsMaxMessageLength"));
    config.setSmsReturnPatternSuccess((String) getFromMap(map, "projectforge", "sms", "returnCodePattern", "success"));
    config.setSmsReturnPatternNumberError((String) getFromMap(map, "projectforge", "sms", "returnCodePattern", "numberError"));
    config.setSmsReturnPatternMessageToLargeError((String) getFromMap(map, "projectforge", "sms", "returnCodePattern", "messageToLargeError"));
    config.setSmsReturnPatternMessageError((String) getFromMap(map, "projectforge", "sms", "returnCodePattern", "messageError"));
    config.setSmsReturnPatternError((String) getFromMap(map, "projectforge", "sms", "returnCodePattern", "error"));
    SmsSender sender = new SmsSender(config);
    sender.send(testNumber, "Hello world! With love by ProjectForge: " + new Date());
  }

  // Helper for getting variables from YAML object map.
  private Object getFromMap(Map<String, Object> map, String... keys) {
    for (String key : keys) {
      Object obj = map.get(key);
      if (obj instanceof Map) {
        map = (Map) obj;
      } else {
        return obj;
      }
    }
    return map;
  }

  private SmsSender.HttpResponseCode testPostCall(String url, String phoneNumber, String message, int fakedReturnCode, String fakedResponseString) throws Exception {
    PostMethod httpMethod = spy(new PostMethod(url));
    SmsSender sender = createSender(url, httpMethod, fakedReturnCode, fakedResponseString);
    SmsSender.HttpResponseCode responseCode = sender.send(phoneNumber, message);
    Assert.assertEquals("smsUser", httpMethod.getParameter("user").getValue());
    Assert.assertEquals("smsPassword", httpMethod.getParameter("password").getValue());
    Assert.assertEquals(message, httpMethod.getParameter("message").getValue());
    Assert.assertEquals(phoneNumber, httpMethod.getParameter("to").getValue());
    verify(sender).createHttpMethod(url);
    return responseCode;
  }

  private SmsSender.HttpResponseCode testGetCall(String url, String phoneNumber, String message, int fakedReturnCode, String fakedResponseString) throws Exception {
    GetMethod httpMethod = spy(new GetMethod(url));
    SmsSender sender = createSender(url, httpMethod, fakedReturnCode, fakedResponseString);
    SmsSender.HttpResponseCode responseCode = sender.send(phoneNumber, message);
    String queryString = httpMethod.getQueryString();
    assertContains(queryString, "user=smsUser");
    assertContains(queryString, "password=smsPassword");
    assertContains(queryString, "message=" + message);
    assertContains(queryString, "to=" + phoneNumber);
    verify(sender).createHttpMethod(url);
    return responseCode;
  }

  private SmsSender createSender(String url, HttpMethod httpMethod, int fakedReturnCode, String fakedResponseString) throws Exception {
    Map<String, String> params = createParams("user", "smsUser", "password", "smsPassword", "message", "#message", "to", "#number");
    SmsSenderConfig config = new SmsSenderConfig();
    config.setHttpMethodType(httpMethod instanceof GetMethod ? "get" : "post")
            .setUrl(url);
    config.setHttpParams(params);
    config.setSmsReturnPatternSuccess("^OK.*");
    config.setSmsReturnPatternError("ERROR");
    config.setSmsReturnPatternMessageError("MESSAGE ERROR");
    config.setSmsReturnPatternMessageToLargeError(".*LARGE.*");
    config.setSmsReturnPatternNumberError("NUMBER FORMAT ERROR");
    SmsSender sender = spy(new SmsSender(config));
    doReturn(httpMethod)
            .when(sender)
            .createHttpMethod(anyString());
    when(httpMethod.getResponseBodyAsString()).thenReturn(fakedResponseString);
    HttpClient httpClient = mock(HttpClient.class);
    doReturn(httpClient)
            .when(sender)
            .createHttpClient();
    when(httpClient.executeMethod(any(HttpMethod.class))).thenReturn(fakedReturnCode);
    return sender;
  }

  private void assertContains(String str, String part) {
    Assert.assertTrue(str.contains(part), "String '" + str + "' must contain '" + part + "'.");
  }

  private Map<String, String> createParams(String... params) {
    Assert.assertTrue(params.length % 2 == 0); // even number.
    Map<String, String> map = new HashMap<>();
    for (int i = 0; i < params.length; i += 2) {
      map.put(params[i], params[i + 1]);
    }
    return map;
  }
}

File: projectforge-business/src/test/kotlin/org/projectforge/messaging/SmsSenderTest.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2022 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.messaging

import org.junit.AfterClass
import org.junit.BeforeClass
import org.junit.Test
import org.junit.jupiter.api.Assertions
import org.mockserver.client.MockServerClient
import org.mockserver.integration.ClientAndServer
import org.mockserver.model.HttpRequest
import org.mockserver.model.HttpResponse
import org.mockserver.model.Parameter
import org.mockserver.model.ParameterBody
import org.projectforge.messaging.SmsSender.HttpResponseCode
import org.projectforge.sms.SmsSenderConfig

class SmsSenderTest {
  private var counter = 0

  @Test
  @Throws(Exception::class)
  fun testSmsService() {
    testCalls(200, "OK, perfect!")
    testCalls(200, "Authentication failed", expectedResponseCode = HttpResponseCode.UNKNOWN_ERROR)
    testCalls(400, "OK", expectedResponseCode = HttpResponseCode.UNKNOWN_ERROR)
  }

  private fun testCalls(
    httpStatus: Int,
    httpResponse: String,
    expectedResponseCode: HttpResponseCode = HttpResponseCode.SUCCESS,
    path: String? = null
  ) {
    val usePath = path ?: "$PATH${++counter}"
    testCall(false, httpStatus, httpResponse, expectedResponseCode, usePath) // GET-call
    testCall(true, httpStatus, httpResponse, expectedResponseCode, usePath)  // POST-call
  }

  private fun testCall(
    post: Boolean,
    httpStatus: Int,
    httpResponse: String,
    expectedResponseCode: HttpResponseCode = HttpResponseCode.SUCCESS,
    path: String = "$PATH${++counter}"
  ) {
    createClientServer(post, httpStatus, httpResponse, path = path)
    val sender = createSender(post, path)
    val responseCode = sender.send(TO, MESSAGE)
    Assertions.assertEquals(expectedResponseCode, responseCode)

  }

  private fun createSender(
    post: Boolean,
    path: String,
    user: String = USER,
    password: String = PASSWORD,
  ): SmsSender {
    val params = createParams("user", user, "password", password, "message", "#message", "to", "#number")
    val config = SmsSenderConfig()
    config.setHttpMethodType(if (post) "post" else "get").url = "http://127.0.0.1:$PORT/$path"
    config.httpParams = params
    config.smsReturnPatternSuccess = "^OK.*"
    config.smsReturnPatternError = "ERROR"
    config.smsReturnPatternMessageError = "MESSAGE ERROR"
    config.smsReturnPatternMessageToLargeError = ".*LARGE.*"
    config.smsReturnPatternNumberError = "NUMBER FORMAT ERROR"
    return SmsSender(config)
  }

  private fun assertContains(str: String, part: String) {
    Assertions.assertTrue(str.contains(part), "String '$str' must contain '$part'.")
  }

  private fun createParams(vararg params: String): Map<String, String> {
    Assertions.assertTrue(params.size % 2 == 0) // even number.
    val map: MutableMap<String, String> = HashMap()
    var i = 0
    while (i < params.size) {
      map[params[i]] = params[i + 1]
      i += 2
    }
    return map
  }

  private fun createClientServer(
    post: Boolean,
    returnHttpCode: Int,
    httpResponse: String,
    message: String = MESSAGE,
    to: String = TO,
    user: String = USER,
    password: String = PASSWORD,
    path: String,
  ): MockServerClient {
    val method = if (post) "POST" else "GET"
    val sc = MockServerClient("127.0.0.1", PORT)
    val request = HttpRequest.request()
      .withMethod(method)
      .withPath("/$path")
    if (post) {
      preparePostRequest(request, message = message, to = to, user = user, password = password)
    } else {
      prepareGetRequest(request, message = message, to = to, user = user, password = password)
    }
    sc.`when`(request)
      .respond(
        HttpResponse.response()
          .withStatusCode(returnHttpCode)
          .withBody(httpResponse)
      )
    return sc
  }

  private fun prepareGetRequest(request: HttpRequest, message: String, to: String, user: String, password: String) {
    request
      .withQueryStringParameter("user", user)
      .withQueryStringParameter("password", password)
      .withQueryStringParameter("message", message)
      .withQueryStringParameter("to", to)
  }

  private fun preparePostRequest(request: HttpRequest, message: String, to: String, user: String, password: String) {
    request.withBody(
      ParameterBody.params(
        Parameter.param("user", user),
        Parameter.param("password", password),
        Parameter.param("message", message),
        Parameter.param("to", to)
      )
    )
  }


  companion object {
    private const val PORT = 65123
    private const val TO = "0123456789"
    private const val MESSAGE = "Hello_world"
    private const val USER = "smsUser"
    private const val PASSWORD = "smsPassword"
    private const val PATH = "send"
    private lateinit var mockServer: ClientAndServer

    @BeforeClass
    @JvmStatic
    fun startServer() {
      mockServer = ClientAndServer.startClientAndServer(PORT)
    }

    @AfterClass
    @JvmStatic
    fun stopServer() {
      mockServer.stop()
    }
  }
}
